### DISCUSSION TITLE
click_threshold in uosc.conf not working?

### DISCUSSION THREAD

**Discussion by hybridearmuff** [tomasklaen/uosc#820] [Q&A]
In the Menu section of uosc's github page it says 
>This is the default pre-configured menu if none is defined in your input.conf, but with added shortcuts. To both pause & move the window with left mouse button, so that you can have the menu on the right one, enable click_threshold in uosc.conf (see default uosc.conf for example/docs).

I'm basically trying to enable this as I want to pause the video, go full screen and move the player with left click and open the menu with right click and I keep changing the click_threshold value in the uosc.conf file in the "scripts-opt" folder but it doesn't seem to be doing anything? All the controls are still exactly the same? I've tried changing it to different values like 1, 200, 500 and 1000 but nothing's changing. I've tried moving the uosc.conf file in the "scripts" folder where uosc is installed besides main.lua but that doesn't work either. I don't have a input.conf file so it couldn't be interfering with that so I'm not sure what I'm doing wrong at this point.

Also, is there some way to change 
>Right click on volume or speed elements to reset them 

to middle click instead? If you're changing the volume or speed with the mouse wheel, it's easier to reset them by just clicking that button instead of moving to another one. I tried checking inside the speed and volume lua files inside the elements folder but I don't think the option is there.

**Comment by tomasklaen** [comment]
> I've tried changing it to different values like 1, 200, 500 and 1000 but nothing's changing.

Does clicking the window not cause play/pause toggling? There is going to be a delay since we need to wait for double click timer to clear, but it should work. We've had some significant event handling changes recently, but I've tested it now and it works like expected.

> Also, is there some way to change right click to reset to middle click instead? 

This is not currently configurable. I don't want to put it on middle click, since you can't really use it on touch pads then. And adding a config for this seems a bit too much. Just use right click :) It's not a big deal.

**Reply by hybridearmuff** [reply]
>Does clicking the window not cause play/pause toggling?

It does but with right click and left click is being used for going full screen and moving the player itself but all that is mpv's default behavior even without uosc. I'm trying to change it to left click to pause/fullscreen/move the window and right to click to open the menu like it says on uosc's github page by enabling click_threshold in uosc.conf but changing its value doesn't seem to be doing anything as all the controls stay exactly the same as if it is still at 0. The only way I can change it to left click is by creating an input.conf file and adding "MBTN_LEFT cycle pause" to it but that creates other problems as it pauses the video when I try to drag the window or double click to go fullscreen and it also doesn't open uosc's menu with right click. At first I thought uosc.conf itself wasn't being recognized by mpv but then I tried changing the value of "speed_step" to 0.5 and that worked as now when I use my mouse wheel to change the speed of the video it changes in increments of 0.5 so it seems like click_threshold is the only option not working. Could it be because I'm using a slightly older version of mpv (version 0.36) as my package manager hasn't updated mpv to its latest version which is 0.37?

>This is not currently configurable. I don't want to put it on middle click, since you can't really use it on touch pads then. And adding a config for this seems a bit too much. Just use right click :) It's not a big deal.

Haha ok but please consider making it configurable in the future if you get the time. Really appreciate all the hard work you do and this amazing piece of software you've created.

**Reply by tomasklaen** [reply]
As documented in `uosc.conf`, `click_threshold` is a value in milliseconds. What it does is execute the `click_command` for clicks shorter than `click_threshold` time (mouse_up_time - mouse_down_time < click_threshold).

So you need to set `click_threshold` to whatever time is most comfortable for you to execute a click. Try `120`.

Then you need to go into your mpv's `input.conf` and set whatever you like to right mouse button, which seem to be uosc menu in your case, so:

```
mbtn_right  script-binding uosc/menu
```

This will give you LMB to pause, double LMB to fullscreen, and RMB for uosc menu.

**Reply by hybridearmuff** [reply]
I added `mbtn_right  script-binding uosc/menu` to input.conf and changed `click_threshold` to 120, left click is still the same and follows standard mpv behavior of moving the player on hold and drag and going fullscreen on double click but it doesn't pause the video and now right click opens the menu instead of cycling pause which is expected. 

I have no way to pause the video now as previously right click was used to cycle pause.

**Reply by tomasklaen** [reply]
Try replacing `uosc/main.lua` with this: [main.zip](https://github.com/tomasklaen/uosc/files/13767322/main.zip) _It just adds logging to click_threshold handling events._

Then launch mpv, click a couple times, and report what it writes in the console.

Default button to toggle console is ` (backtick).

**Reply by hybridearmuff** [reply]
I replaced the original main.lua with the one you provided in main.zip and it breaks uosc for me, as in uosc doesn't funtion at all and I get the default mpv player layout. Checking in the console it says 

```
[uosc] script-opts/uosc.conf:160 unknown key 'text_width_estimation', ignoring 
[uosc] 
[uosc] stack traceback:
[uosc]  /home/xxx/.config/mpv/scripts/uosc/main.lua:132: in main chunk
[uosc]  [C]: at 0x55eda193b780
[uosc]  [C]: at 0x55eda193bf30
[uosc] Lua error: /home/xxx/.config/mpv/scripts/uosc/main.lua:132: module 'lib/char_conv' not found:
[uosc]  no field package.preload['lib/char_conv']
[uosc]  no file '/home/xxx/.config/mpv/scripts/uosc/lib/char_conv.lua'
[uosc]  no file '/usr/share/luajit-2.1/lib/char_conv.lua'
[uosc]  no file '/usr/local/share/lua/5.1/lib/char_conv.lua'
[uosc]  no file '/usr/local/share/lua/5.1/lib/char_conv/init.lua'
[uosc]  no file '/usr/share/lua/5.1/lib/char_conv.lua'
[uosc]  no file '/usr/share/lua/5.1/lib/char_conv/init.lua'
[uosc]  no file '/usr/local/lib/lua/5.1/lib/char_conv.so'
[uosc]  no file '/usr/lib/lua/5.1/lib/char_conv.so'
[uosc]  no file '/usr/local/lib/lua/5.1/loadall.so'
```

and clicking doesn't show anything in the console either. These errors don't show up when using the original main.lua that was in the uosc folder.

Sorry accidentally didn't add the first line from console. Edited now.

**Reply by tomasklaen** [reply]
Ah, you're not on latest main branch. You can instead replace the `-- Click detection` block in your `main.lua` with this one:

```lua
-- Click detection
if options.click_threshold > 0 then
	-- Executes custom command for clicks shorter than `options.click_threshold`
	-- while filtering out double clicks.
	local click_time = options.click_threshold / 1000
	local doubleclick_time = mp.get_property_native('input-doubleclick-time') / 1000
	local last_down, last_up = 0, 0
	local click_timer = mp.add_timeout(math.max(click_time, doubleclick_time), function()
		local delta = last_up - last_down
		print('delta:', delta, ', click_time:', click_time)
		if delta > 0 and delta < click_time and delta > 0.02 then mp.command(options.click_command) end
	end)
	click_timer:kill()
	mp.set_key_bindings({{'mbtn_left',
		function() print('up') last_up = mp.get_time() end,
		function()
			print('down')
			last_down = mp.get_time()
			if click_timer:is_enabled() then click_timer:kill() else click_timer:resume() end
		end,
	}}, 'mouse_movement', 'force')
	mp.enable_key_bindings('mouse_movement', 'allow-vo-dragging+allow-hide-cursor')
end
```

It just adds a couple print statements.

**Reply by hybridearmuff** [reply]
Here's what I got after clicking a bunch of times and then double clicking a few times
```
 (+) Video --vid=1 (*) (h264 1920x1080 23.976fps)
 (+) Audio --aid=1 --alang=eng (*) (aac 2ch 44100Hz)
AO: [pipewire] 44100Hz stereo 2ch floatp
VO: [gpu-next] 1920x1080 yuv420p
AV: 00:00:01 / 00:03:07 (1%) A-V:  0.004 DS: 3.120/0
[uosc] down 
[uosc] up 
AV: 00:00:01 / 00:03:07 (1%) A-V:  0.004 DS: 2.969/0
[uosc] delta: 0.00014200000000031 , click_time: 0.12 
AV: 00:00:01 / 00:03:07 (1%) A-V:  0.004 DS: 2.895/0
[uosc] down 
[uosc] up 
AV: 00:00:01 / 00:03:07 (1%) A-V:  0.004 DS: 2.844/0
[uosc] delta: 5.6999999999974e-05 , click_time: 0.12 
AV: 00:00:02 / 00:03:07 (1%) A-V:  0.004 DS: 2.759/0
[uosc] down 
[uosc] up 
AV: 00:00:02 / 00:03:07 (1%) A-V:  0.004 DS: 2.738/0
[uosc] delta: 2.300000000055e-05 , click_time: 0.12 
AV: 00:00:03 / 00:03:07 (2%) A-V:  0.004 DS: 2.692/0
[uosc] down 
[uosc] up 
AV: 00:00:03 / 00:03:07 (2%) A-V:  0.004 DS: 2.682/0
[uosc] delta: 2.300000000055e-05 , click_time: 0.12 
AV: 00:00:04 / 00:03:07 (2%) A-V:  0.004 DS: 2.653/0
[uosc] down 
[uosc] up 
AV: 00:00:04 / 00:03:07 (2%) A-V:  0.004 DS: 2.660/0
[uosc] down 
[uosc] up 
AV: 00:00:04 / 00:03:07 (3%) A-V:  0.005 DS: 2.500/0
[uosc] down 
AV: 00:00:04 / 00:03:07 (3%) A-V:  0.005 DS: 2.500/0
[uosc] up 
AV: 00:00:05 / 00:03:07 (3%) A-V:  0.005 DS: 2.500/0
[uosc] down 
AV: 00:00:05 / 00:03:07 (3%) A-V:  0.005 DS: 2.500/0
[uosc] up 
AV: 00:00:05 / 00:03:07 (3%) A-V:  0.005 DS: 2.490/1
[uosc] down 
[uosc] up 
AV: 00:00:05 / 00:03:07 (3%) A-V:  0.005 DS: 2.490/1
[uosc] down 
[uosc] up 
AV: 00:00:06 / 00:03:07 (3%) A-V:  0.004 DS: 2.490/1
[uosc] down 
AV: 00:00:06 / 00:03:07 (3%) A-V:  0.004 DS: 2.490/1
[uosc] up 
AV: 00:00:06 / 00:03:07 (3%) A-V:  0.004 DS: 2.490/1
[uosc] down 
AV: 00:00:06 / 00:03:07 (4%) A-V:  0.014 DS: 2.470/2
[uosc] up 
AV: 00:00:09 / 00:03:07 (5%) A-V:  0.016 DS: 2.480/2

Exiting... (Quit)
```
Seems like it's working.

**Reply by christoph-heinrich** [reply]
Inb4 they say they're on wayland :upside_down_face: (because deltas are very small)

Doesn't work for me either, I'll investigate.

**Reply by christoph-heinrich** [reply]
In windowed mode it doesn't work for me because of the 0.02s minimum delta, which is good because of wayland.

In fullscreen it works only when no elements are visible. That's because when elements things become visible they register event handlers and we start listening to left clicks in input.lua, which shadows the keybinding for click_threshold.

We need to either forward left click events that don't hit any elements to a click_threshold handler, or unconditionally listen to left clicks in cursor.lua handle that in there.

**Reply by tomasklaen** [reply]
@christoph-heinrich I've tried moving the click detection to `cursor:zone()` handlers, but I actually can't figure out how are we preventing window dragging in those? In `cursor:decide_keybids()`, the `'allow-vo-dragging+allow-hide-cursor'` flags are always being passed to the `mbtn_left` bindings, but the dragging is still enabled above no zone, and prevented above zones like speed/volume sliders. How are we doing this? :)

**Reply by christoph-heinrich** [reply]
@tomasklaen https://github.com/tomasklaen/uosc/pull/708#issuecomment-1765322627

**Comment by tomasklaen** [comment]
So #831 fixes this problem, but some side effects still need to be hammered out before merging.
